### PR TITLE
CapabilityRegistry: addDON

### PR DESCRIPTION
**Pull request by DeividasK**

### CODE REVIEW

File: contracts/src/v0.8/keystone/CapabilityRegistry.sol [hunk 1]
```diff
@@ -71,6 +74,35 @@ contract CapabilityRegistry is OwnerIsCreator, TypeAndVersionInterface {
...
+    bytes32 capabilityId;
+    // This configuration is expected to be decodeable on-chain (not enforced).
+    // Any configuration that does not need to be verified on chain should live
+    // in the `offchainConfig`. This configuration can be updated together with
+    // DON nodes.
+    bytes onchainConfig;
+    // This allows fine-grained configuration of capabilities across DON
+    // instances. This configuration can be updated together with DON nodes.
+    bytes offchainConfig;
+  }
```
Comment (DeividasK) [inline contracts/src/v0.8/keystone/CapabilityRegistry.sol]: This will be removed.

File: contracts/src/v0.8/keystone/CapabilityRegistry.sol [hunk 2]
```diff
@@ -71,6 +74,35 @@ contract CapabilityRegistry is OwnerIsCreator, TypeAndVersionInterface {
...
+  // the same capabilities.
+  struct DON {
+    // Computed. Auto-increment.
+    uint32 id;
+    // Whether the DON accepts external capability requests.
+    bool isPublic;
+    // A set of p2pIds of nodes that belong to this DON. A node (the same
+    // p2pId) can belong to multiple DONs.
+    bytes32[] nodes;
+    CapabilityConfiguration[] capabilitiesConfigurations;
```
Comment (DeividasK) [inline contracts/src/v0.8/keystone/CapabilityRegistry.sol]: This will change to mapping(bytes32 hashedCapabilityId => bytes config)

File: contracts/src/v0.8/keystone/CapabilityRegistry.sol [hunk 3]
```diff
@@ -77,6 +80,33 @@ contract CapabilityRegistry is OwnerIsCreator, TypeAndVersionInterface {
...
 
+  /// @notice CapabilityConfiguration is a struct that holds the capability configuration
+  /// for a specific DON
+  struct CapabilityConfiguration {
+    /// @notice The capability Id
+    bytes32 capabilityId;
+    /// @notice The capability config specific to a DON.  This will be decoded
+    /// offchain
+    bytes config;
+  }
```
Comment (DeividasK) [inline contracts/src/v0.8/keystone/CapabilityRegistry.sol:91]: Should be removed?

File: contracts/src/v0.8/keystone/CapabilityRegistry.sol [hunk 4]
```diff
@@ -124,10 +159,16 @@ contract CapabilityRegistry is OwnerIsCreator, TypeAndVersionInterface {
   /// exists.
   error CapabilityAlreadyExists();
 
+  /// @notice This error is thrown when trying to add a capability configuration
+  /// for a capability that was already configured on a DON
+  /// @param donId The ID of the DON that the capability was configured for
+  /// @param capabilityId The ID of the capability that was configured
+  error DonCapabilityExists(uint32 donId, bytes32 capabilityId);
```
Comment (DeividasK) [inline contracts/src/v0.8/keystone/CapabilityRegistry.sol]:
```suggestion
  error DuplicateDONCapability(uint32 donId, bytes32 capabilityId);
```
WDYT?

File: contracts/src/v0.8/keystone/CapabilityRegistry.sol [hunk 5]
```diff
@@ -124,10 +159,16 @@ contract CapabilityRegistry is OwnerIsCreator, TypeAndVersionInterface {
...
+
   /// @notice This error is thrown when a capability with the provided hashed ID is
   /// not found.
-  /// @param hashedCapabilityId The hashed ID used for the lookup.
-  error CapabilityDoesNotExist(bytes32 hashedCapabilityId);
+  /// @param capabilityId The hashed ID used for the lookup.
```
Comment (DeividasK) [inline contracts/src/v0.8/keystone/CapabilityRegistry.sol]: Why the change?

File: contracts/src/v0.8/keystone/CapabilityRegistry.sol [hunk 6]
```diff
@@ -406,4 +454,59 @@ contract CapabilityRegistry is OwnerIsCreator, TypeAndVersionInterface {
...
+    bool isPublic
+  ) external onlyOwner {
+    uint32 id = s_donId;
+
+    s_dons[id].id = id;
+    s_dons[id].isPublic = isPublic;
+
+    for (uint256 i; i < capabilityConfigurations.length; ++i) {
+      CapabilityConfiguration calldata configuration = capabilityConfigurations[i];
+      bytes32 capabilityIds = configuration.capabilityId;
```
Comment (DeividasK) [inline contracts/src/v0.8/keystone/CapabilityRegistry.sol]:
```suggestion
      bytes32 capabilityId = configuration.capabilityId;
```

File: contracts/src/v0.8/keystone/CapabilityRegistry.sol [hunk 7]
```diff
@@ -406,4 +454,59 @@ contract CapabilityRegistry is OwnerIsCreator, TypeAndVersionInterface {
...
+
+      if (s_capabilities[capabilityIds].version == bytes32("")) revert CapabilityDoesNotExist(capabilityIds);
```
Comment (DeividasK) [inline contracts/src/v0.8/keystone/CapabilityRegistry.sol]:
```suggestion
      if (!s_hashedCapabilityIds.contains(capabilityIds)) revert CapabilityDoesNotExist(capabilityIds);
```

File: contracts/src/v0.8/keystone/CapabilityRegistry.sol [hunk 8]
```diff
@@ -406,4 +454,59 @@ contract CapabilityRegistry is OwnerIsCreator, TypeAndVersionInterface {
...
+
+      if (s_dons[id].capabilityIds.contains(capabilityIds)) revert DonCapabilityExists(id, capabilityIds);
+
+      s_dons[id].capabilityIds.add(capabilityIds);
+      s_dons[id].capabilityConfigs[capabilityIds] = configuration.config;
+    }
+
```
Comment (DeividasK) [inline contracts/src/v0.8/keystone/CapabilityRegistry.sol:544]: TODO: Verify capability is not deprecated https://chainlink-core.slack.com/archives/C06F7S6T2BU/p1714585501653989

File: contracts/src/v0.8/keystone/CapabilityRegistry.sol (hunk 4 above)
Comment (aiagentdev-1990) [reply contracts/src/v0.8/keystone/CapabilityRegistry.sol]: updated

File: contracts/src/v0.8/keystone/CapabilityRegistry.sol (hunk 5 above)
Comment (aiagentdev-1990) [reply contracts/src/v0.8/keystone/CapabilityRegistry.sol]: This was accidental.  I've reverted

File: contracts/src/v0.8/keystone/CapabilityRegistry.sol (hunk 7 above)
Comment (aiagentdev-1990) [reply contracts/src/v0.8/keystone/CapabilityRegistry.sol]: updated

File: contracts/src/v0.8/keystone/CapabilityRegistry.sol [hunk 9]
```diff
@@ -406,4 +459,68 @@ contract CapabilityRegistry is OwnerIsCreator, TypeAndVersionInterface {
...
+      if (!s_hashedCapabilityIds.contains(capabilityId)) revert CapabilityDoesNotExist(capabilityId);
+
+      if (s_dons[id].capabilityIds.contains(capabilityId)) revert DuplicateDONCapability(id, capabilityId);
+
+      s_dons[id].capabilityIds.add(capabilityId);
+      s_dons[id].capabilityConfigs[capabilityId] = configuration.config;
+    }
+
+    // TODO:  Verify that the nodes added support the relevant capability
```
Comment (aiagentdev-1990) [inline contracts/src/v0.8/keystone/CapabilityRegistry.sol]: This is currently not trivial as the `Node` struct just has a `bytes32[]` field for the supported capability IDs.  We have 2 options to make this easier.

1.  We update this to an `EnumerableSet.Bytes32Set` so that we can check that a node supports a capability in O(1).  The downside with this is that we'd have to update the function signatures for `addNodes` and `updateNodes` as the `Node` struct will now have a nested mapping

2.  We define a separate `mapping(bytes32 nodeP2PId => EnumerableSet.Bytes32Set capabilityIds)` to avoid the problem above.  This does duplicate data though.

@DeividasK

File: contracts/src/v0.8/keystone/CapabilityRegistry.sol [hunk 10]
```diff
@@ -124,6 +155,24 @@ contract CapabilityRegistry is OwnerIsCreator, TypeAndVersionInterface {
   /// exists.
   error CapabilityAlreadyExists();
 
+  /// @notice This error is thrown when trying to add a node to a DON where
+  /// the node does not support the capability
+  /// @param donId The ID of the DON
+  /// @param nodeP2PId The P2P ID of the node
+  /// @param capabilityId The ID of the capability
+  error NodeDoesNotSupportDONCapability(uint32 donId, bytes32 nodeP2PId, bytes32 capabilityId);
```
Comment (DeividasK) [inline contracts/src/v0.8/keystone/CapabilityRegistry.sol]:
```suggestion
  error NodeDoesNotSupportCapability(bytes32 nodeP2PId, bytes32 hashedCapabilityId);
```
I think DON is not relevant in this error. WDYT?

Comment (aiagentdev-1990) [reply contracts/src/v0.8/keystone/CapabilityRegistry.sol]: updated

File: contracts/src/v0.8/keystone/CapabilityRegistry.sol [hunk 11]
```diff
@@ -33,7 +33,9 @@ contract CapabilityRegistry is OwnerIsCreator, TypeAndVersionInterface {
     address signer;
   }
 
-  struct NodeCapability {
+  struct Node {
+    /// @notice The node's parameters
+    NodeParams params;
```
Comment (DeividasK) [inline contracts/src/v0.8/keystone/CapabilityRegistry.sol]: Does this correctly batch `address signer` and `uint32 configCount`?

Comment (aiagentdev-1990) [reply contracts/src/v0.8/keystone/CapabilityRegistry.sol]: Packed these into a single struct

File: contracts/src/v0.8/keystone/CapabilityRegistry.sol [hunk 12]
```diff
@@ -406,4 +504,73 @@ contract CapabilityRegistry is OwnerIsCreator, TypeAndVersionInterface {
...
+      s_dons[id].capabilityIds.add(capabilityId);
+      s_dons[id].capabilityConfigs[capabilityId] = configuration.config;
+    }
+
+    for (uint256 i; i < nodes.length; ++i) {
+      bytes32 nodeP2PId = nodes[i];
+      if (s_dons[id].nodes.contains(nodeP2PId)) revert DuplicateDONNode(id, nodeP2PId);
+
+      s_dons[id].nodes.add(nodeP2PId);
+    }
```
Comment (DeividasK) [inline contracts/src/v0.8/keystone/CapabilityRegistry.sol]: This should be done before we add capabilities. Let's write a test to catch a bug where capability is added, and then a node is added without the support for the said capability.

Comment (DeividasK) [reply contracts/src/v0.8/keystone/CapabilityRegistry.sol]: Nvm, this is fine.

File: contracts/src/v0.8/keystone/CapabilityRegistry.sol [hunk 13]
```diff
@@ -296,38 +378,54 @@ contract CapabilityRegistry is OwnerIsCreator, TypeAndVersionInterface {
...
-    return s_nodes[p2pId];
+  /// @return NodeParams The node data
+  function getNode(bytes32 p2pId) external view returns (NodeParams memory, bytes32[] memory) {
+    return (
+      NodeParams({
+        nodeOperatorId: s_nodes[p2pId].nodeOperatorId,
+        p2pId: s_nodes[p2pId].p2pId,
+        signer: s_nodes[p2pId].signer
+      }),
+      s_nodes[p2pId].supportedCapabilityIds[s_nodes[p2pId].configCount].values()
```
Comment (DeividasK) [inline contracts/src/v0.8/keystone/CapabilityRegistry.sol]: Should also include `configCount`, just in case 🤷‍♂️

File: contracts/src/v0.8/keystone/CapabilityRegistry.sol [hunk 14]
```diff
@@ -19,19 +22,37 @@ contract CapabilityRegistry is OwnerIsCreator, TypeAndVersionInterface {
...
-    uint256 nodeOperatorId;
+    uint32 nodeOperatorId;
+    /// @notice The signer address for application-layer message verification.
+    address signer;
     /// @notice This is an Ed25519 public key that is used to identify a node.
     /// This key is guaranteed to be unique in the CapabilityRegistry. It is
     /// used to identify a node in the the P2P network.
     bytes32 p2pId;
+    /// @notice The list of supported hashed capability IDs supported
+    /// by the node
```
Comment (DeividasK) [inline contracts/src/v0.8/keystone/CapabilityRegistry.sol]:
```suggestion
    /// @notice The list of hashed capability IDs supported by the node
```

File: contracts/src/v0.8/keystone/CapabilityRegistry.sol [hunk 15]
```diff
@@ -246,29 +323,34 @@ contract CapabilityRegistry is OwnerIsCreator, TypeAndVersionInterface {
...
       if (nodeExists || bytes32(node.p2pId) == bytes32("")) revert InvalidNodeP2PId(node.p2pId);
 
       if (node.signer == address(0)) revert InvalidNodeSigner();
 
-      if (node.supportedHashedCapabilityIds.length == 0)
-        revert InvalidNodeCapabilities(node.supportedHashedCapabilityIds);
+      bytes32[] memory capabilityIds = node.hashedCapabilityIds;
+      if (capabilityIds.length == 0) revert InvalidNodeCapabilities(capabilityIds);
+
+      s_nodes[node.p2pId].configCount++;
```
Comment (DeividasK) [inline contracts/src/v0.8/keystone/CapabilityRegistry.sol]: Nit: ++

File: contracts/src/v0.8/keystone/CapabilityRegistry.sol [hunk 16]
```diff
@@ -406,4 +500,81 @@ contract CapabilityRegistry is OwnerIsCreator, TypeAndVersionInterface {
...
+
+    emit DONAdded(id, isPublic);
+  }
+
+  /// @notice Get's a DON's data
+  /// @param donId The DON's ID
+  /// @return uint32 The DON ID
+  /// @return bool True if the DON is public
+  /// @return bytes32[] The list of node P2P IDs that are in the DON
+  /// @return CapabilityConfiguration[] The list of capability configurations supported by the DON
```
Comment (DeividasK) [inline contracts/src/v0.8/keystone/CapabilityRegistry.sol]:
```suggestion
```suggestion
  /// @notice Gets DON's data
  /// @param donId The DON ID
  /// @return bool True if the DON is public
  /// @return bytes32[] The list of node P2P IDs that are in the DON
  /// @return CapabilityConfiguration[] The list of capability configurations supported by the DON
```

File: contracts/src/v0.8/keystone/test/CapabilityRegistry_AddDONTest.t.sol [hunk 17]
```diff
@@ -0,0 +1,183 @@
...
+        CapabilityRegistry.NodeDoesNotSupportCapability.selector,
+        P2P_ID_TWO,
+        s_capabilityWithConfigurationContractId
+      )
+    );
+    s_capabilityRegistry.addDON(nodes, capabilityConfigs, true);
+  }
+
+  function test_RevertWhen_CapabilityDoesNotExist() public {
+    changePrank(ADMIN);
```
Comment (DeividasK) [inline contracts/src/v0.8/keystone/test/CapabilityRegistry_AddDONTest.t.sol]: Could be moved to setup?

**Comment by DeividasK** [comment]
@cds95 took this over. I approve this PR.

**Review by aiagentdev-1990 — DISMISSED** [review]

File: contracts/src/v0.8/keystone/CapabilityRegistry.sol [hunk 18]
```diff
@@ -19,19 +22,36 @@ contract CapabilityRegistry is OwnerIsCreator, TypeAndVersionInterface {
     string name;
   }
 
-  struct Node {
+  struct NodeParams {
     /// @notice The id of the node operator that manages this node
-    uint256 nodeOperatorId;
+    uint32 nodeOperatorId;
+    /// @notice The signer address for application-layer message verification.
+    address signer;
```
Comment (makramkd) [inline contracts/src/v0.8/keystone/CapabilityRegistry.sol:29]: Is this the onchain signing key? `address` only works for EVM - `bytes` is more general (maybe `bytes32` also works, I don't think there are any chains that have public keys > 32 bytes atm)

File: contracts/src/v0.8/keystone/CapabilityRegistry.sol [hunk 19]
```diff
@@ -7,6 +7,9 @@ import {IERC165} from "../vendor/openzeppelin-solidity/v4.8.3/contracts/interfac
 import {EnumerableSet} from "../vendor/openzeppelin-solidity/v4.8.3/contracts/utils/structs/EnumerableSet.sol";
 import {ICapabilityConfiguration} from "./interfaces/ICapabilityConfiguration.sol";
 
+// CapabilityRegistry is used to manage Nodes (including their links to Node
+// Operators), Capabilities, and DONs (Decentralized Oracle Networks) which are
+// sets of nodes that support those Capabilities.
```
Comment (makramkd) [inline contracts/src/v0.8/keystone/CapabilityRegistry.sol]: Could be a docstring with the @ notice stuff.

File: contracts/src/v0.8/keystone/CapabilityRegistry.sol [hunk 20]
```diff
@@ -19,19 +22,36 @@ contract CapabilityRegistry is OwnerIsCreator, TypeAndVersionInterface {
...
-    bytes32[] supportedHashedCapabilityIds;
+    /// @notice This is an Ed25519 public key that is used to identify a node.
+    /// This key is guaranteed to be unique in the CapabilityRegistry. It is
+    /// used to identify a node in the the P2P network.
+    bytes32 p2pId;
+    /// @notice The node's supported capabilities
+    /// @dev This is stored as a map so that we can easily update to a set of
+    /// new capabilities by incrementing the configCount and creating a
+    /// new set of supported capability IDs
+    mapping(uint32 configCount => EnumerableSet.Bytes32Set capabilityId) supportedCapabilityIds;
```
Comment (makramkd) [inline contracts/src/v0.8/keystone/CapabilityRegistry.sol:54]: Is there a use case for seeing the capabilities supported by a node at a previous config count?

Comment (makramkd) [reply contracts/src/v0.8/keystone/CapabilityRegistry.sol:54]: Asking because we only seem to return the latest - is this just less gassy?

File: contracts/src/v0.8/keystone/CapabilityRegistry.sol (hunk 18 above)
Comment (makramkd) [reply contracts/src/v0.8/keystone/CapabilityRegistry.sol:29]: If this is the offchain signing public key, its too small (Ed25519 used there)

File: contracts/src/v0.8/keystone/CapabilityRegistry.sol [hunk 21]
```diff
@@ -406,4 +499,81 @@ contract CapabilityRegistry is OwnerIsCreator, TypeAndVersionInterface {
...
+
+    return (s_dons[donId].id, s_dons[donId].isPublic, s_dons[donId].nodes.values(), capabilityConfigurations);
+  }
+
+  /// @notice Returns the DON specific configuration for a capability
+  /// @param donId The DON's ID
+  /// @param capabilityId The Capability ID
+  /// @return bytes The DON specific configuration for the capability
+  function getDONCapabilityConfig(uint32 donId, bytes32 capabilityId) external view returns (bytes memory) {
+    return s_dons[donId].capabilityConfigs[capabilityId];
```
Comment (makramkd) [inline contracts/src/v0.8/keystone/CapabilityRegistry.sol:577]: Should this also return the configuration contract's `getCapabilityConfiguration(donId)` as well?

File: contracts/src/v0.8/keystone/CapabilityRegistry.sol (hunk 20 above)
Comment (aiagentdev-1990) [reply contracts/src/v0.8/keystone/CapabilityRegistry.sol:54]: Yes we basically need a quick way to track the latest supported capability IDs after `updateNodes` is called.  The alternative would have been to loop through all the capability IDs and remove them from the set one by one and adding the new capability IDs.  Adding the `configCount` allows us to just read the latest supported IDs.

File: contracts/src/v0.8/keystone/CapabilityRegistry.sol (hunk 18 above)
Comment (aiagentdev-1990) [reply contracts/src/v0.8/keystone/CapabilityRegistry.sol:29]: I've created a follow up ticket to address this

File: contracts/src/v0.8/keystone/CapabilityRegistry.sol (hunk 19 above)
Comment (aiagentdev-1990) [reply contracts/src/v0.8/keystone/CapabilityRegistry.sol]: added

**Review by makramkd — APPROVED** [review]

**Review by aiagentdev-1990 — APPROVED** [review]
